Stake pool/liquid deposits (rebased on new master)

jon-chuang opened the pull request:
@billythedummy

This PR adds liquid deposits as well as deposit fees.

The PR builds on the merged PR https://github.com/solana-labs/solana-program-library/pull/2107

With these changes, we believe stake pools will be much safer and easier to use.

joncinque commented on stake-pool/program/tests/deposit_sol.rs:
```diff
@@ -0,0 +1,467 @@
+}
+
+#[tokio::test]
+async fn success_with_referral_fee() {
```
Great tests!

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -311,8 +317,15 @@ fn command_create_pool(
     ];
-    unique_signers!(initialize_signers);
-    initialize_transaction.sign(&initialize_signers, recent_blockhash);
+    if let Some(stake_deposit_authority) = stake_deposit_authority {
+        let mut initialize_signers = initialize_signers.clone();
+        initialize_signers.push(&stake_deposit_authority);
+        unique_signers!(initialize_signers);
+        initialize_transaction.sign(&initialize_signers, recent_blockhash);
+    } else {
+        unique_signers!(initialize_signers);
+        initialize_transaction.sign(&initialize_signers, recent_blockhash);
+    }
```
Oops, thanks for fixing this!

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -347,6 +360,15 @@ fn command_vsa_create(
     Ok(())
 }
 
+fn command_show(config: &Config, stake_pool_address: &Pubkey) -> CommandResult {
+    let stake_pool = get_stake_pool(&config.rpc_client, stake_pool_address)?;
+    println!(
+        "Stake pool at address {}:\n{:#?}",
+        stake_pool_address, stake_pool
+    );
+    Ok(())
+}
+
```
Can we remove this and add any additional required info to `list`?

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -567,11 +589,12 @@ fn add_associated_token_account(
     stake_pool_address: &Pubkey,
     stake: &Pubkey,
     token_receiver: &Option<Pubkey>,
+    referrer: &Option<Pubkey>,
```
nit: just to disambiguate, can we call this `referrer_token_account`?  Just this interface makes it seem like it will derive an ATA from this pubkey

jon-chuang replied:
good point. In general, referrer is a fixed entity, else the token receiver. Perhaps `token_receiver` should also be renamed `pool_token_receiver_account`.

joncinque replied:
Yes please, great idea!

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -660,8 +688,125 @@ fn command_deposit(
+        command_update(config, stake_pool_address, false, false)?;
+    }
+
+    let amount = spl_token::ui_amount_to_amount(amount, 9);
```
Since you're using sol here, best to use `native_token::sol_to_lamports`

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -660,8 +688,125 @@ fn command_deposit(
+    if from_balance < amount {
+        return Err(format!(
+            "Not enough SOL to deposit into pool: {}.\nMaximum deposit amount is {} SOL.",
+            spl_token::amount_to_ui_amount(amount, 9),
+            spl_token::amount_to_ui_amount(from_balance, 9)
```
Here you can use `Sol(lamports)` to get the pretty sign, and not have to write `SOL` in the message

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -660,8 +688,125 @@ fn command_deposit(
+    stake_pool_address: &Pubkey,
+    from: &Option<Keypair>,
+    token_receiver: &Option<Pubkey>,
+    referrer: &Option<Pubkey>,
```
nit: can we call this `referrer_token_account`?

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -660,8 +688,125 @@ fn command_deposit(
+        config.sol_depositor.as_ref()
+    {
+        let expected_sol_deposit_authority = stake_pool.sol_deposit_authority.ok_or_else(|| {
+            "SOL deposit authority specified in config but stake Pool has none".to_string()
```
```suggestion
            "SOL deposit authority specified in arguments but stake pool has none".to_string()
```

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -696,7 +841,7 @@ fn command_list(config: &Config, stake_pool_address: &Pubkey) -> CommandResult {
         println!("Manager: {}", stake_pool.manager);
         println!("Staker: {}", stake_pool.staker);
-        println!("Depositor: {}", stake_pool.deposit_authority);
+        println!("Depositor: {}", stake_pool.stake_deposit_authority);
```
Can we print the sol deposit authority if it exists? And also the deposit, withdrawal, and referral fees?

jon-chuang replied:
yep, didn't realise this functionality existed!

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -660,8 +688,125 @@ fn command_deposit(
+
+    let stake_pool = get_stake_pool(&config.rpc_client, stake_pool_address)?;
+
+    let mut instructions: Vec<Instruction> = vec![];
```
For added safety, let's craft this transaction as:

* create ephemeral keypair
* transfer `amount` from `from` to the ephemeral keypair
* pass the ephemeral keypair to `sol_deposit` and sign with it

this way, people can feel better and know that the maximum amount at risk is however much is transferred to the ephemeral keypair, and *not* everything in their account

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -1581,8 +1826,8 @@ fn main() {
-        .subcommand(SubCommand::with_name("deposit")
-            .about("Add stake account to the stake pool")
+        .subcommand(SubCommand::with_name("deposit-stake")
+            .about("Deposit fully-activated stake account into the stake pool in exchange for pool tokens")
```
This is a nice clarification. We've had people confused about using the term `fully-activated` in the past, so just a slight simplification:
```suggestion
            .about("Deposit active stake account into the stake pool in exchange for pool tokens")
```

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -1611,6 +1856,61 @@ fn main() {
+                    .value_name("ADDRESS")
+                    .takes_value(true)
+                    .help("Account to receive the referral fees for deposits. \
+                          Defaults to the pool manager."),
```
```suggestion
                          Defaults to the token receiver."),
```

billythedummy replied:
I believe the referrer is supposed to default to the pool manager if none is provided

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -1611,6 +1856,61 @@ fn main() {
+                    .validator(is_pubkey)
+                    .value_name("ADDRESS")
+                    .takes_value(true)
+                    .help("Account to receive the referral fees for deposits. \
```
```suggestion
                    .help("Pool token account to receive the referral fees for deposits. \
```

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -1611,6 +1856,61 @@ fn main() {
+                    .validator(is_keypair)
+                    .value_name("KEYPAIR")
+                    .takes_value(true)
+                    .help("Account to deposit SOL from. \
```
nit: we use this in other parts of the CLI, like `create-stake-account`
```suggestion
                    .help("Source account of funds. \
```

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -1611,6 +1856,61 @@ fn main() {
+            .arg(
+                Arg::with_name("from")
+                    .long("from")
+                    .validator(is_keypair)
```
If people want to use this with ledger, we'll need to support `is_keypair_or_ask_keyword` -- what do you think?

jon-chuang replied:
This doesn't really grant support for ledger, does it?

That really requires remote wallet

jon-chuang replied:
Anw, I've added the ask keyword in case ppl want it

joncinque replied:
We're using remote wallet already, right?

jon-chuang replied:
Then isn't what you want, is_signer?

jon-chuang replied:
Oh, I see, you always ask the config for this.

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -1611,6 +1856,61 @@ fn main() {
+                    .validator(is_pubkey)
+                    .value_name("REFERRER_TOKEN_ADDRESS")
+                    .takes_value(true)
+                    .help("Account to receive the referral fees for deposits. \
+                          Defaults to the pool manager."),
```
Just thinking, if someone is using this from the CLI, how about they get the referral instead?
```suggestion
                    .help("Pool token account to receive the referral fees for deposits. \
                          Defaults to the token receiver."),
```

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -1648,8 +1948,8 @@ fn main() {
-        .subcommand(SubCommand::with_name("withdraw")
-            .about("Withdraw amount from the stake pool")
+        .subcommand(SubCommand::with_name("withdraw-stake")
+            .about("Withdraw amount from the stake pool into designated stake account")
```
The destination stake account isn't required, so this addition could confuse people

jon-chuang replied:
That's true. It would be nice to have the destination be an existing mergeable user stake account, however. 

The ephemeral stake account would be deleted once merged with the desination, in a single atomic transaction.

joncinque replied:
Yeah that's been on my mental list for a bit... it would be a great feature

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -660,8 +688,125 @@ fn command_deposit(
+        &mut total_rent_free_balances,
+    ));
+
+    let referrer = referrer.unwrap_or(stake_pool.manager_fee_account);
```
Kindness to users?
```suggestion
    let referrer = referrer.unwrap_or(token_receiver);
```

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -621,36 +644,41 @@ fn command_deposit(
         &mut total_rent_free_balances,
     ));
 
+    let referrer = referrer.unwrap_or(stake_pool.manager_fee_account);
```
We :heart: CLIs?
```suggestion
    let referrer = referrer.unwrap_or(token_receiver);
```

jon-chuang replied:
Ok, user deserves credit for using the cli :P

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -1432,6 +1665,18 @@ fn main() {
                     .help("The validator vote account that this stake will be delegated to"),
             )
         )
+        .subcommand(SubCommand::with_name("show")
+            .about("Show the state pool state")
+            .arg(
+                Arg::with_name("pool")
+                    .index(1)
+                    .validator(is_pubkey)
+                    .value_name("POOL_ADDRESS")
+                    .takes_value(true)
+                    .required(true)
+                    .help("Stake pool address"),
+            )
+        )
```
Let's remove this too

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -1758,8 +2058,36 @@ fn main() {
+                    .help("Stake pool address."),
+            )
+            .arg(
+                Arg::with_name("new_sol_deposit_authority")
+                    .index(2)
+                    .validator(|x| {
+                        #[allow(clippy::cmp_owned)]
+                        let default = if x == String::from("none") {
+                            Ok(())
+                        } else {
+                            Err(String::from("Not a pubkey or 'none'"))
+                        };
+                        is_pubkey(x).or(default)
+                    })
+                    .value_name("ADDRESS_OR_NONE")
+                    .takes_value(true)
+                    .help("'none', or a public key for the new stake pool sol deposit authority."),
```
How about following the convention from the preferred validator CLI?  There, you specify `--unset` if you want to unset or `--vote-account <PUBKEY_HERE>` to specify the preferred validator vote account.  I'm not sure which is better, but we should at least be consistent

jon-chuang replied:
Ok, I prefer `--set-to-none`, and have the two options `conflicts_with` one another

joncinque replied:
Sure! Check out where `--unset` is being used, and rename all of those too then.  It works exactly in that way

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -1769,9 +2097,17 @@ fn main() {
+            .arg(Arg::with_name("fee_type")
+                .index(2)
+                .value_name("FEE_TYPE")
+                .possible_values(&["pool","deposit", "withdrawal"]) // PreferredValidatorType enum
```
There's three different names for the original fee at this point: management, pool, and epoch.  `pool` is probably too vague, so I'm happy with either `management` or `epoch`, with a slight preference toward `epoch`.  Either way, let's keep it consistent

jon-chuang replied:
ok, epoch it is.

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -1884,18 +2222,21 @@ fn main() {
             let denominator = value_t_or_exit!(arg_matches, "fee_denominator", u64);
             let w_numerator = value_t!(arg_matches, "withdrawal_fee_numerator", u64);
             let w_denominator = value_t!(arg_matches, "withdrawal_fee_denominator", u64);
+            let d_numerator = value_t!(arg_matches, "withdrawal_fee_numerator", u64);
```
```suggestion
            let d_numerator = value_t!(arg_matches, "deposit_fee_numerator", u64);
```

jon-chuang replied:
oops, thanks!

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -1884,18 +2222,21 @@ fn main() {
             let w_numerator = value_t!(arg_matches, "withdrawal_fee_numerator", u64);
             let w_denominator = value_t!(arg_matches, "withdrawal_fee_denominator", u64);
+            let d_numerator = value_t!(arg_matches, "withdrawal_fee_numerator", u64);
+            let d_denominator = value_t!(arg_matches, "withdrawal_fee_denominator", u64);
```
```suggestion
            let d_denominator = value_t!(arg_matches, "deposit_fee_denominator", u64);
```

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -1904,12 +2245,21 @@ fn main() {
                 reserve_keypair,
             )
         }
+        ("show", Some(arg_matches)) => {
+            let stake_pool_address = pubkey_of(arg_matches, "pool").unwrap();
+            command_show(&config, &stake_pool_address)
+        }
```
Let's remove this too

joncinque reviewed (commented):
Really great work here! It was a lot to add all of these, so I really appreciate the time that you spent on this.  A lot of the comments are pretty minor, but the biggest one will be the separation of sol deposit fees and stake deposit fees.

jon-chuang commented on stake-pool/program/src/processor.rs:
```diff
@@ -813,20 +830,13 @@ impl Processor {
+        {
             msg!(
-                "Error: attempting to add stake with {} lamports, must have {} lamports",
+                "Error: attempting to add (stake: {}, delegation: {}), below minimum",
```
unfortunately, message can only take up to 4 arguments.

The reason for combining the error is that I encounter a stackoverflow for the instruction unless combining this particular if/else.

When trying to format the minimum into a single tuple, I also encountered the stackoverflow error.

I'll try again with a different build config.

joncinque replied:
ah crap, that really stinks.  Since they're hard-coded values, it isn't a huge deal, you can let this go

jon-chuang commented on stake-pool/program/src/processor.rs:
```diff
@@ -1762,6 +1774,14 @@ impl Processor {
+            return Err(StakePoolError::InvalidFeeAccount.into());
+        }
+
+        if stake_pool.token_program_id != *referrer_fee_info.owner {
+            return Err(StakePoolError::InvalidFeeAccount.into());
+        }
```
one less if-else condition, I like it

jon-chuang commented on stake-pool/program/src/state.rs:
```diff
@@ -115,12 +115,17 @@ pub struct StakePool {
+
+    /// Toggles whether the `DepositSol` instruction requires a signature from
+    /// the `deposit_authority`
+    pub sol_deposit_authority: Option<Pubkey>,
```
Yeah, I was thinking the same. More flexibility would definitely help.

jon-chuang commented on stake-pool/program/tests/huge_pool.rs:
```diff
@@ -32,7 +32,7 @@ use {
 };
 
-const HUGE_POOL_SIZE: u32 = 4_000;
+const HUGE_POOL_SIZE: u32 = 3_950;
```
I'm not sure this time... 4000 is definitely the nicer number, but, hopefully that won't matter in due time with message processor/VM improvements. 

Still, I'll investigate a bit.

joncinque replied:
It's not a huge deal considering we're adding quite a bit of new logic to do the extra mints on deposit, don't spend too much time on it, I'd say.

jon-chuang commented on stake-pool/program/tests/set_withdrawal_fee.rs:
```diff
@@ -321,42 +321,6 @@ async fn fail_high_withdrawal_fee_increase_from_0() {
     }
 }
 
-#[tokio::test]
```
It's actually a duplicate of fail_high_withdrawal_fee

joncinque replied:
Ah ok, disregard then, thanks!

jon-chuang commented:
@joncinque how do you feel about refactoring the deposit authority so that it always is controlled by the program, and then make it possible to toggle the deposit authority on/off? The struct would change so that deposit authority is set as the program's permanently, and there is a new field `manager_stake_deposit_authority: Option<Pubkey>`, and the old one is renamed `stake_account_deposit_authority`.

It's very inflexible the way it is currently, as it's really all-or-nothing at the create pool stage.

We'd like the authority to gate deposits into the pool. 

Not sure if it will break compatibiltility with deployed testnet pools.

To retain backwards compatibility, we can not enforce that the `stake_account_deposit_authority` be program derived. But we recommend that the gating stake deposits be done via `manager_stake_deposit_authority` going forward.

joncinque commented:
We can go that way.  Otherwise what if we control the ability to deposit through the fee?  If it's set to 100%, that's like saying "pool is private".  Then we don't need to change the deposit authority.

jon-chuang commented:
@joncinque , I think that approach only works for private pools, but not if you want to gate deposits on a per-validator basis. 

In which case, we're better off with a separate authority, so if that makes sense to you, we'll go ahead and implement that.

Is there still a use case for initializing with a non PDA deposit authority then? If not, should we remove that from the initialization?

joncinque commented:
@jon-chuang let me make sure i understand what you're suggesting

* `manager_deposit_stake_authority`: new field that can be toggled by the manager, forces deposits to be signed by this account
* `stake_account_deposit_authority`: always have this field as a PDA

Then on deposit, we just add a check for the signature if `manager_deposit_stake_authority` is set.  Before deposit, we always set the withdraw / staking authority to `stake_account_deposit_authority`.  Is that right?

jon-chuang commented:
> @jon-chuang let me make sure i understand what you're suggesting
> 
> * `manager_deposit_stake_authority`: new field that can be toggled by the manager, forces deposits to be signed by this account
> * `stake_account_deposit_authority`: always have this field as a PDA
> 
> Then on deposit, we just add a check for the signature if `manager_deposit_stake_authority` is set. Before deposit, we always set the withdraw / staking authority to `stake_account_deposit_authority`. Is that right?

Yes, that's precisely right.

joncinque commented:
All right, sounds good to me! I see what you mean about existing stake pools... let's not worry too much about it -- there's mainly the foundation stake pool, and i can always hack together an instruction to upgrade it once we're ready

billythedummy commented on stake-pool/program/src/instruction.rs:
```diff
@@ -960,11 +1104,14 @@ pub fn withdraw(
         AccountMeta::new_readonly(sysvar::clock::id(), false),
         AccountMeta::new_readonly(*token_program_id, false),
         AccountMeta::new_readonly(stake_program::id(), false),
+        AccountMeta::new_readonly(system_program::id(), false),
```
yeah, seems like it. We'll remove this

billythedummy commented on stake-pool/program/src/instruction.rs:
```diff
@@ -925,11 +989,91 @@ pub fn deposit_with_authority(
+    token_program_id: &Pubkey,
+    amount: u64,
+) -> Vec<Instruction> {
+    let dummy_sol_deposit_authority = Pubkey::default();
```
i think we want to keep those fields separated for the use case where the `deposit_sol_authority` is an off-chain server or an on-chain PDA that users must go through for deposits.

jon-chuang commented:
Actually, @joncinque , realistically, the why shouldn't the program simply use the `withdraw_authority` to sign off on everything instead of requiring the `deposit_authority`? We could get rid of the `stake_account_deposit_authority` entirely

jon-chuang commented:
CI is misbehaving, I decided to rebase again.

jon-chuang commented:
@joncinque since we now have `deposit_stake_fee` and `deposit_sol_fee`, does this mean that we also need a `stake_referral_fee` and `sol_referral_fee`? It seems like way too much to have to manage. Could we unify the referral fees in this case?

joncinque commented:
ah, that seems to be the best solution!  and we'll still keep the "default" PDA as an option, correct?

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -696,10 +852,20 @@ fn command_list(config: &Config, stake_pool_address: &Pubkey) -> CommandResult {
         println!("Withdraw Authority: {}", pool_withdraw_authority);
         println!("Pool Token Mint: {}", stake_pool.pool_mint);
         println!("Fee Account: {}", stake_pool.manager_fee_account);
+        println!("Epoch Fee: {}", stake_pool.fee);
```
This is now being printed twice -- could you remove the other print?

billythedummy replied:
oh i didn't realize that block was there already, i think we can standardize that if/else block for all the fee types. It provides more info for the user

joncinque replied:
Either way works for me :smile:

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -659,9 +679,142 @@ fn command_deposit(
+    let mut total_rent_free_balances: u64 = 0;
+
+    // Create the ephemeral SOL account
+    instructions.push(system_instruction::create_account(
```
nit: you can just do a `transfer`, since the `allocate` + `assign` part of `create_account` is unnecessary

joncinque commented on stake-pool/program/src/lib.rs:
```diff
@@ -115,4 +117,4 @@ pub fn find_transient_stake_program_address(
 }
 
-solana_program::declare_id!("SPoo1xuN9wGpxNjGnPNbRPtpQ7mHgKM8d9BeFC549Jy");
+solana_program::declare_id!("5ocnV1qiCgaQR8Jb8xWnVbApfaygJ8tNoZfgPwsgx9kx");
```
:eyes:

billythedummy replied:
oops haha

joncinque commented on stake-pool/program/src/processor.rs:
```diff
@@ -2224,6 +2410,36 @@ impl Processor {
+        stake_pool.check_manager(manager_info)?;
+        if for_stake_deposit {
+            stake_pool.stake_deposit_authority = new_sol_deposit_authority.unwrap_or(
+                find_deposit_authority_program_address(&crate::id(), stake_pool_info.key).0,
```
Use the passed `program_id` here instead of `crate::id()`, and you won't have to change the crate id

joncinque commented on stake-pool/program/src/instruction.rs:
```diff
@@ -315,10 +323,81 @@ pub enum StakePoolInstruction {
         fee: Fee,
     },
+
+    ///  (Manager only) Update DepositStake referral fee
```
```suggestion
    ///  (Manager only) Update DepositSol referral fee
```

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -2018,17 +2472,33 @@ fn main() {
                 numerator,
             };
-            command_set_fee(&config, &stake_pool_address, new_fee)
+            match arg_matches.value_of("fee_type").unwrap() {
```
This change-set is already getting pretty massive, but we could reduce the number of instructions and simplify the code quite a bit if we actually add this as an enum, `FeeType`.  That can be passed into generic `SetFee` and `SetReferralFee`, which would reduce the number of instructions from 5 to 2.  It would also get rid of that `for_stake_deposit` bool, which isn't very clear unfortunately.  What do you think?

billythedummy replied:
What do you think of combining everything into 1 enum?

```
enum FeeType {
  ReferreralSol(u8),
  ReferralStake(u8),
  Epoch(Fee),
  Withdrawal(Fee),
  Deposit(Fee),
}
```

joncinque replied:
Yeah sure, that works!

jon-chuang commented:
> ah, that seems to be the best solution! and we'll still keep the "default" PDA as an option, correct?

Yes, that's correct. Not initializing or unsetting the stake_deposit_authority turns it back to the PDA via the priorly defined seeds.

joncinque commented:
Is this good to go from your side? I pushed some changes which will require a rebase unfortunately.  Don't worry about adapting to the changes too much, I mainly added a check on deposit / withdraw to make sure that a person is actually getting back more than 0 tokens, and then changed the rounding for the deposit / withdraw calcs.  Once you rebase and merge, then I can add the calcs in

jon-chuang commented:
To not let matters get out of hand, given that this is a 2.9K LOC PR, here is a summary of changes:


### Instructions:
Modified instructions
1. Renaming: `Deposit` -> `DepositStake`, `Withdraw`->`WithdrawStake`. 
2. Unified update fee instruction. `SetFee { Fee }` -> `SetFee { FeeType }` where `FeeType = enum { ${type}Fee(Fee) }`
3. Initialize also initializes with `stake_deposit_fee` and `stake_referral_fee`. Not sure if we want to get rid of that, we can initialize then update immediately through the cli.
4. `DepositStake` also accepts manager and referrer fee accounts to credit `deposit_stake_fee` and corresponding `stake_referral_fee`.

New instructions
1. `DepositSol`
2. `SetSolDepositAuthority`
3. `SetStakeDepositAuthority`. Leverages previously defined PDA.

Removed instructions:
1. `SetWithdrawalFee` (unified into `SetFee`)

TODO:
1. Could probably unify `deposit_sol_with_authority` by passing an option instead of having two instruction constructors.
2. Not sure if we want to maintain separate instruction constructors for each `FeeType`, or to simply expose the generic versions (right now, we expose both).
3. Not sure if we want to unify `SetStakeDepositAuthority` and `SetSolDepositAuthority` into single instruction, also using enums. Currently, the two instructions call the same `process_set_deposit_authority`.

### Processor changes (most crucial security-wise)
1. Apply fees to `DepositStake`. Additional checks: manager fee account matches `StakePool` struct's.
2. Add `DepositSol` instruction, basically same as `DepositStake` except for source of lamports (activated stake accounts' staked SOL + rent-free reserve v.s. signing source account). The major difference is `DepositSol` checks the system program (for carrying out SOL transfer), does not ingest or check the validator list account.
3. Unified `process_set_deposit_authority`.

Potential Security issues:
1. whether or not additional fee calculations introduce weird integer error bugs. (to a certain degree, no, as the errors are < 1 lamport per fee application).
2. Whether or not any mistakes were made in calculating input SOL v.s. output pool tokens (probably not, due to common use of previously defined `stake_pool.calc_pool_tokens_for_deposit`.

TODO:
1. Maybe we should add asserts to ensure that `pool_tokens_user + pool_tokens_manager_deposit_fee + pool_tokens_referral_fee == new_pool_tokens`.

### Struct changes:
1. `pub sol_deposit_authority: Option<Pubkey>`
2. `pub sol_deposit_fee: Fee`
3. `pub sol_referral_fee: u8`

### Tests:
We've added tests for all of the changes, including:
- set deposit authority
- set deposit fee 
- set referral fee
- deposit sol
- new test for deposit stake (with referral fee/invalid referrer)

###  Cli changes: 
1. create pool uses `stake_deposit_authority` keypair to initialize. 
2. new commands: 
a. `deposit_sol`, 
b. `set_deposit_authority` (for both deposit sol and deposit stake, via `for_stake_deposit` boolean).
c. monomorphized as `set-${type}-deposit-authority`.
d. Unified `set-stake-referral-fee` for both stake and sol deposits.
3. modified commands: `deposit-stake` (+ referrer account, + manager fee). command `list` (display new fields). Unified `set_fee` across 4 `Fee` kinds to reduce code bloat. 

TODO:
1. `--from` can be a hardware wallet (nice to have, future PR).

jon-chuang commented:
@joncinque I think I'm happy with this PR.

In terms of future changes we think would be good (nice to have, but not critical):
1. Multiple transient stake accounts (up to about 8). Iterate over a range in `[0, MAX_TRANSIENT_ACCS)` which we add to the transient account seed. I think we all agree that a single transient account is neither necessary nor ideal. To 

I saw you mentioning that the stake pool code is undergoing another round of audits, will these changes make it in time? We'd like to wrap things up on the stake pool program front for our first deployment to mainnet, hopefully in about 1 week's time. Perhaps if things can move fast enough we may depend on the solana-deployed version of the program after all, and it can be made to eventually be immutable, just like the Token program is. 

In terms of immutability, I'm just a bit worried about slashing/changes to the stake program.

joncinque commented:
Great work!
